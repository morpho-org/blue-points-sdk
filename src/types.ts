import { Address as ViemAddress, Hex as ViemHex } from "viem";

// Re export the types from viem in order to not make viem as a peer dependency
export type Address = ViemAddress;
export type Hex = ViemHex;

export enum PositionType {
  SUPPLY,
  BORROW,
  COLLATERAL,
}

export type EntityId = string;

export interface Entity {
  id: EntityId;
}

/**
 * MarketPoints is the points representation of the Market entity.
 * It contains aggregated values that are derivated from the positions.
 */
export interface MarketPoints extends Entity {
  /* The ID of the market. This is the same as the onchain id, lowercased. */
  id: Hex;

  /** The total supply points in the market.
   *
   * __Note__: This should be equal to the sum of all supply points in the market positions.
   */
  totalSupplyPoints: bigint;
  /** The total borrow points in the market.
   *
   * __Note__: This should be equal to the sum of all borrow points in the market positions.
   */
  totalBorrowPoints: bigint;
  /** The total collateral in the market.
   *
   * __Note__: This should be equal to the sum of all collateral in the market positions.
   */
  totalCollateralPoints: bigint;
}

/**
 * Market is the entity that represents a market in the system.
 * It extends the points and add indexes + timestamps, used to compute a market snapshot.
 */
export interface Market extends MarketPoints {
  totalSupplyShares: bigint;
  totalBorrowShares: bigint;
  totalCollateral: bigint;

  lastUpdate: bigint;
}

/**
 * PositionPoints is the points representation of the Position entity.
 * It contains all the points & points of a user position in a given market
 */
export interface PositionPoints extends Entity {
  /** The Checksummed address of the user. */
  user: Address;
  /** The ID of the market. This is the same as the onchain id, lowercased. */
  market: Hex;

  /** The total supply points in the position. */
  supplyPoints: bigint;
  /** The total borrow points in the position. */
  borrowPoints: bigint;
  /** The total collateral in the position. */
  collateralPoints: bigint;
}

/**
 * Position is the entity that represents a user position in the system.
 * It extends the points and add indexes + timestamps, used to compute a position snapshot.
 */
export interface Position extends PositionPoints {
  /** The total supply shares of the position, at the time used to retrieve the state */
  supplyShares: bigint;
  /** The total borrow shares of the position, at the time used to retrieve the state */
  borrowShares: bigint;
  /** The total collateral of the position, at the time used to retrieve the state */
  collateral: bigint;

  /** The last update timestamp of the position. */
  lastUpdate: bigint;
}

/**
 * Tx is the entity that represents a MetaMorpho in the system.
 *
 * It contains aggregated values that are derivated from the vault positions.
 */
export interface MetaMorphoPoints extends Entity {
  /** The Checksummed address of the vault. */
  id: Address;

  /** The total supply points in the vault.
   *
   * __Note__: This should be equal to the sum of all supply points in the vault positions.
   */
  totalPoints: bigint;

  /** The total supply points in the vault.
   *
   * __Note__: This number is computed from indexes, that can lead to rounding errors.
   */
}

/**
 * MetaMorpho is the entity that represents a MetaMorpho in the system.
 *
 * It extends the points and add indexes + timestamps, used to compute a MetaMorpho snapshot.
 */
export interface MetaMorpho extends MetaMorphoPoints {
  /** The total shares of the vault, at the time used to retrieve the state */
  totalShares: bigint;

  /** The last update timestamp of the vault. */
  lastUpdate: bigint;
}

/**
 * MetaMorphoPositionPoints is the points representation of the MetaMorphoPosition entity.
 */
export interface MetaMorphoPositionPoints extends Entity {
  /** The Checksummed address of the user. */
  user: Address;
  /** The Checksummed address of the vault. */
  metaMorpho: Address;

  /** The total supply points in the position. */
  supplyPoints: bigint;
}

/**
 * MetaMorphoPosition is the entity that represents a user position in the system.
 * It extends the points and add indexes + timestamps, used to compute a position snapshot.
 */
export interface MetaMorphoPosition extends MetaMorphoPositionPoints {
  /** The total shares of the position, at the time used to retrieve the state */
  shares: bigint;

  /** The last update timestamp of the position. */
  lastUpdate: bigint;
}

export interface Tx extends Entity {
  timestamp: bigint;

  txHash: Hex;
  txIndex: bigint;
  logIndex: bigint;
  blockNumber: bigint;
}

export interface MetaMorphoTx extends Tx {
  metaMorpho: Address;
  market: Hex;
  user: Address;

  position: EntityId; // MetaMorphoPosition id;
  shares: bigint;
  timestamp: bigint;
  type: PositionType;
}

export interface MorphoTx extends Tx {
  type: PositionType;
  user: Address;
  market: Address;
  shares: bigint;
}
