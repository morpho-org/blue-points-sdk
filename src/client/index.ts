export type { State, PointsState } from "./state";
export { clonePointsState } from "./state";
export { default as PointsClient } from "./client";
