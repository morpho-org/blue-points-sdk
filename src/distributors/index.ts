export * from "./distributors";
export * from "./redistributor";
export * from "./stateDiff";
export * from "./utils";
export * from "./constants";
