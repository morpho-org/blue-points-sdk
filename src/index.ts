export { Address, Hex } from "./types";
export * as BluePoints from "./types";

export * from "./distributors";
export * from "./stateManager";
export * from "./loaders";
export * from "./programs";
export * from "./modules";
export * from "./checkers";

export const config = {
  cacheEnabled: true,
};

export const setConfig = (newConfig: Partial<typeof config>) => {
  if (Object.isFrozen(config)) throw new Error("Config is frozen");
  Object.assign(config, {
    ...config,
    ...newConfig,
  });
};
export const getConfig = () => {
  Object.freeze(config);
  return config;
};
