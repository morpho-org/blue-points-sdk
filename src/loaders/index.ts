export * from "./fullLoader";
export * from "./userLoader";
