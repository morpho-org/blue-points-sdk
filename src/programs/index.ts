export * from "./programs";
export * from "./merkleTree";
