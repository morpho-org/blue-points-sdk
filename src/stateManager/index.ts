export type { State, PointsState } from "./state";
export { clonePointsState } from "./state";
