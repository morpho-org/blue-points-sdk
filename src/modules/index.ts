import { PointsState } from "../client";

export * from "./blacklisting";

export interface Module {
  handle: (state: PointsState) => PointsState;
}
