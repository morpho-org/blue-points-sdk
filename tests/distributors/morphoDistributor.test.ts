import { Address, getAddress, Hex } from "viem";

import {
  computeMarketPoints,
  computePositionPoints,
  getPositionId,
  handleMorphoTx,
  initPosition,
  initPositionPoints,
} from "../../src/distributors/morphoDistributor";
import { Market, MorphoTx, Position, PositionType } from "../../src/types";

import { state } from "./mock";

describe("Morpho distributor", () => {
  it("should getPositionId return a concatenation of the market id & the user address", () => {
    const marketId: Hex = "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b";
    const user: Address = getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327");
    const positionId = getPositionId(marketId, user);
    expect(positionId).toEqual(
      "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b51382c7e0ff4dd26683b3356f4fd1320f9cf3327"
    );
  });
  it("should initPositionPoints create the correct entity", () => {
    const marketId: Hex = "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b";
    const user: Address = getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327");
    const position = initPositionPoints(marketId, user);

    expect(position).toMatchSnapshot();
  });

  it("should initPosition create the correct entity", () => {
    const marketId: Hex = "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b";
    const user: Address = getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327");
    const position = initPosition(marketId, user);

    expect(position).toMatchSnapshot();
  });

  it("should computeMarketPoints update the suupply points correctly", () => {
    const market = state.markets[
      "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b"
    ] as Market;
    const timestamp = 1712113403n + 10n;

    const editedMarket = computeMarketPoints(market, timestamp);

    expect(editedMarket.totalSupplyPoints).toEqual(
      36000000000000000000000000n + 10n * market.totalSupplyShares
    );
    expect(editedMarket.totalBorrowPoints).toEqual(0n);
    expect(editedMarket.totalCollateralPoints).toEqual(19680000n + 10n * market.totalCollateral);
    expect(editedMarket.lastUpdate).toEqual(timestamp);

    // check if the original market is not mutated
    expect(market.totalSupplyPoints).toEqual(36000000000000000000000000n);
    expect(market.totalBorrowPoints).toEqual(0n);
    expect(market.totalCollateralPoints).toEqual(19680000n);
    expect(market.lastUpdate).toEqual(1712113403n);
  });

  it("should computeMarketPoints throws if there is a future lastUpdate", () => {
    const market = state.markets[
      "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b"
    ] as Market;
    const timestamp = 1712113403n - 10n;

    expect(() => computeMarketPoints(market, timestamp)).toThrowError(
      `Market ${market.id} has a future lastUpdate`
    );
  });

  it("should computePositionPoints update the points correctly", () => {
    const position = state.positions[
      getPositionId(
        "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
        getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327")
      )
    ] as Position;
    const timestamp = 1712671931n + 10n;

    const editedPosition = computePositionPoints(position, timestamp);

    expect(editedPosition.supplyPoints).toEqual(0n + 10n * position.supplyShares);
    expect(editedPosition.borrowPoints).toEqual(
      6304686594341171851252854600n + 10n * position.borrowShares
    );
    expect(editedPosition.collateralPoints).toEqual(
      2464285725637273525921295808n + 10n * position.collateral
    );
    expect(editedPosition.lastUpdate).toEqual(timestamp);

    // check if the original position is not mutated
    expect(position.supplyPoints).toEqual(0n);
    expect(position.borrowPoints).toEqual(6304686594341171851252854600n);
    expect(position.collateralPoints).toEqual(2464285725637273525921295808n);
    expect(position.lastUpdate).toEqual(1712671931n);
  });

  it("should computePositionPoints throws if there is a future lastUpdate", () => {
    const position = state.positions[
      getPositionId(
        "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
        getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327")
      )
    ] as Position;

    const timestamp = 1712671931n - 10n;

    expect(() => computePositionPoints(position, timestamp)).toThrowError(
      `Position ${position.id} has a future lastUpdate`
    );
  });

  it("should handleMorphoTx update the state correctly for a supply transaction", () => {
    const initialState = { ...state };

    const tx: MorphoTx = {
      market: "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
      user: getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327"),
      shares: 100n,
      timestamp: 1712671931n + 10n,
      type: PositionType.SUPPLY,
      id: "0x1",
      blockNumber: 1n,
      logIndex: 1n,
      txHash: "0x1",
      txIndex: 1n,
    };

    const newState = handleMorphoTx(initialState, tx);

    const initialMarket = initialState.markets[tx.market] as Market;
    const initialPosition = initialState.positions[getPositionId(tx.market, tx.user)] as Position;

    const finalMarket = newState.markets[tx.market] as Market;
    const finalPosition = newState.positions[getPositionId(tx.market, tx.user)] as Position;

    const deltaTMarket = tx.timestamp - initialMarket.lastUpdate;

    expect(finalMarket.totalSupplyPoints).toEqual(
      initialMarket.totalSupplyPoints + deltaTMarket * initialMarket.totalSupplyShares
    );
    expect(finalMarket.totalBorrowPoints).toEqual(
      initialMarket.totalBorrowPoints + deltaTMarket * initialMarket.totalBorrowShares
    );
    expect(finalMarket.totalCollateralPoints).toEqual(
      initialMarket.totalCollateralPoints + deltaTMarket * initialMarket.totalCollateral
    );
    expect(finalMarket.lastUpdate).toEqual(tx.timestamp);

    expect(finalPosition.supplyPoints).toEqual(
      initialPosition.supplyPoints + 10n * initialPosition.supplyShares
    );
    expect(finalPosition.borrowPoints).toEqual(
      initialPosition.borrowPoints + 10n * initialPosition.borrowShares
    );
    expect(finalPosition.collateralPoints).toEqual(
      initialPosition.collateralPoints + 10n * initialPosition.collateral
    );

    // accounting
    expect(finalPosition.supplyShares).toEqual(initialPosition.supplyShares + tx.shares);
    expect(finalPosition.borrowShares).toEqual(initialPosition.borrowShares);
    expect(finalPosition.collateral).toEqual(initialPosition.collateral);

    expect(finalMarket.totalSupplyShares).toEqual(initialMarket.totalSupplyShares + tx.shares);
    expect(finalMarket.totalBorrowShares).toEqual(initialMarket.totalBorrowShares);
    expect(finalMarket.totalCollateral).toEqual(initialMarket.totalCollateral);
  });

  it("should handleMorphoTx update the state correctly for a new user", () => {
    const initialState = { ...state };

    const tx: MorphoTx = {
      market: "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
      user: getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3330"),
      shares: 100n,
      timestamp: 1712671931n + 10n,
      type: PositionType.SUPPLY,
      id: "0x1",
      blockNumber: 1n,
      logIndex: 1n,
      txHash: "0x1",
      txIndex: 1n,
    };

    const newState = handleMorphoTx(initialState, tx);

    const initialMarket = initialState.markets[tx.market] as Market;

    const finalMarket = newState.markets[tx.market] as Market;
    const finalPosition = newState.positions[getPositionId(tx.market, tx.user)] as Position;

    const deltaTMarket = tx.timestamp - initialMarket.lastUpdate;

    expect(finalMarket.totalSupplyPoints).toEqual(
      initialMarket.totalSupplyPoints + deltaTMarket * initialMarket.totalSupplyShares
    );
    expect(finalMarket.totalBorrowPoints).toEqual(
      initialMarket.totalBorrowPoints + deltaTMarket * initialMarket.totalBorrowShares
    );
    expect(finalMarket.totalCollateralPoints).toEqual(
      initialMarket.totalCollateralPoints + deltaTMarket * initialMarket.totalCollateral
    );
    expect(finalMarket.lastUpdate).toEqual(tx.timestamp);

    expect(finalPosition.supplyPoints).toEqual(0n);
    expect(finalPosition.borrowPoints).toEqual(0n);
    expect(finalPosition.collateralPoints).toEqual(0n);

    // accounting
    expect(finalPosition.supplyShares).toEqual(tx.shares);
    expect(finalPosition.borrowShares).toEqual(0n);
    expect(finalPosition.collateral).toEqual(0n);

    expect(finalMarket.totalSupplyShares).toEqual(initialMarket.totalSupplyShares + tx.shares);
    expect(finalMarket.totalBorrowShares).toEqual(initialMarket.totalBorrowShares);
    expect(finalMarket.totalCollateral).toEqual(initialMarket.totalCollateral);
  });

  it("should handleMorphoTx update the state correctly for a borrow transaction", () => {
    const initialState = { ...state };

    const tx: MorphoTx = {
      market: "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
      user: getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327"),
      shares: 100n,
      timestamp: 1712671931n + 10n,
      type: PositionType.BORROW,
      id: "0x1",
      blockNumber: 1n,
      logIndex: 1n,
      txHash: "0x1",
      txIndex: 1n,
    };

    const newState = handleMorphoTx(initialState, tx);

    const initialMarket = initialState.markets[tx.market] as Market;
    const initialPosition = initialState.positions[getPositionId(tx.market, tx.user)] as Position;

    const finalMarket = newState.markets[tx.market] as Market;
    const finalPosition = newState.positions[getPositionId(tx.market, tx.user)] as Position;

    const deltaTMarket = tx.timestamp - initialMarket.lastUpdate;

    expect(finalMarket.totalSupplyPoints).toEqual(
      initialMarket.totalSupplyPoints + deltaTMarket * initialMarket.totalSupplyShares
    );
    expect(finalMarket.totalBorrowPoints).toEqual(
      initialMarket.totalBorrowPoints + deltaTMarket * initialMarket.totalBorrowShares
    );
    expect(finalMarket.totalCollateralPoints).toEqual(
      initialMarket.totalCollateralPoints + deltaTMarket * initialMarket.totalCollateral
    );
    expect(finalMarket.lastUpdate).toEqual(tx.timestamp);

    expect(finalPosition.supplyPoints).toEqual(
      initialPosition.supplyPoints + 10n * initialPosition.supplyShares
    );
    expect(finalPosition.borrowPoints).toEqual(
      initialPosition.borrowPoints + 10n * initialPosition.borrowShares
    );
    expect(finalPosition.collateralPoints).toEqual(
      initialPosition.collateralPoints + 10n * initialPosition.collateral
    );

    // accounting
    expect(finalPosition.supplyShares).toEqual(initialPosition.supplyShares);
    expect(finalPosition.borrowShares).toEqual(initialPosition.borrowShares + tx.shares);
    expect(finalPosition.collateral).toEqual(initialPosition.collateral);

    expect(finalMarket.totalSupplyShares).toEqual(initialMarket.totalSupplyShares);
    expect(finalMarket.totalBorrowShares).toEqual(initialMarket.totalBorrowShares + tx.shares);
    expect(finalMarket.totalCollateral).toEqual(initialMarket.totalCollateral);
  });

  it("should handleMorphoTx update the state correctly for a collateral transaction", () => {
    const initialState = { ...state };

    const tx: MorphoTx = {
      market: "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
      user: getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327"),
      shares: 100n,
      timestamp: 1712671931n + 10n,
      type: PositionType.COLLATERAL,
      id: "0x1",
      blockNumber: 1n,
      logIndex: 1n,
      txHash: "0x1",
      txIndex: 1n,
    };

    const newState = handleMorphoTx(initialState, tx);

    const initialMarket = initialState.markets[tx.market] as Market;
    const initialPosition = initialState.positions[getPositionId(tx.market, tx.user)] as Position;

    const finalMarket = newState.markets[tx.market] as Market;
    const finalPosition = newState.positions[getPositionId(tx.market, tx.user)] as Position;

    const deltaTMarket = tx.timestamp - initialMarket.lastUpdate;

    expect(finalMarket.totalSupplyPoints).toEqual(
      initialMarket.totalSupplyPoints + deltaTMarket * initialMarket.totalSupplyShares
    );
    expect(finalMarket.totalBorrowPoints).toEqual(
      initialMarket.totalBorrowPoints + deltaTMarket * initialMarket.totalBorrowShares
    );
    expect(finalMarket.totalCollateralPoints).toEqual(
      initialMarket.totalCollateralPoints + deltaTMarket * initialMarket.totalCollateral
    );
    expect(finalMarket.lastUpdate).toEqual(tx.timestamp);

    expect(finalPosition.supplyPoints).toEqual(
      initialPosition.supplyPoints + 10n * initialPosition.supplyShares
    );
    expect(finalPosition.borrowPoints).toEqual(
      initialPosition.borrowPoints + 10n * initialPosition.borrowShares
    );
    expect(finalPosition.collateralPoints).toEqual(
      initialPosition.collateralPoints + 10n * initialPosition.collateral
    );

    // accounting
    expect(finalPosition.supplyShares).toEqual(initialPosition.supplyShares);
    expect(finalPosition.borrowShares).toEqual(initialPosition.borrowShares);
    expect(finalPosition.collateral).toEqual(initialPosition.collateral + tx.shares);

    expect(finalMarket.totalSupplyShares).toEqual(initialMarket.totalSupplyShares);
    expect(finalMarket.totalBorrowShares).toEqual(initialMarket.totalBorrowShares);
    expect(finalMarket.totalCollateral).toEqual(initialMarket.totalCollateral + tx.shares);
  });

  it("should handleMorphoTx throw if the market is not found", () => {
    const tx: MorphoTx = {
      market: "0xAAA",
      user: getAddress("0x51382c7e0ff4dd26683b3356f4fd1320f9cf3327"),
      shares: 100n,
      timestamp: 1712671931n + 10n,
      type: PositionType.COLLATERAL,
      id: "0x1",
      blockNumber: 1n,
      logIndex: 1n,
      txHash: "0x1",
      txIndex: 1n,
    };
    expect(() => handleMorphoTx(state, tx)).toThrowError(`Market ${tx.market} not found`);
  });
});
