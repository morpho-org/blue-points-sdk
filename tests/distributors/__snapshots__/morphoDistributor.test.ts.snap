// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Morpho distributor should initPosition create the correct entity 1`] = `
{
  "borrowPoints": 0n,
  "borrowShares": 0n,
  "collateral": 0n,
  "collateralPoints": 0n,
  "id": "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b51382c7e0ff4dd26683b3356f4fd1320f9cf3327",
  "lastUpdate": 0n,
  "market": "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
  "supplyPoints": 0n,
  "supplyShares": 0n,
  "user": "0x51382C7E0ff4dD26683B3356f4FD1320f9cf3327",
}
`;

exports[`Morpho distributor should initPositionPoints create the correct entity 1`] = `
{
  "borrowPoints": 0n,
  "collateralPoints": 0n,
  "id": "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b51382c7e0ff4dd26683b3356f4fd1320f9cf3327",
  "market": "0x06cb6aaee2279b46185dc2c8c107b4a56ff6550ea86063ec011fa4a52920841b",
  "supplyPoints": 0n,
  "user": "0x51382C7E0ff4dD26683B3356f4FD1320f9cf3327",
}
`;
